fiona/_vectorized.pyx

```diff
@@ -0,0 +1,188 @@
+from .ogrext cimport Session, _deleteOgrFeature
+from .ogrext import FIELD_TYPES, FIELD_TYPES_MAP, OGRERR_NONE
+from ._shim cimport *
+from libc.stdlib cimport malloc, free
+from fiona.rfc3339 import FionaDateType, FionaDateTimeType, FionaTimeType
+
+import logging
+from six import text_type
+import datetime
+
+import numpy as np
+cimport numpy as np
+
+log = logging.getLogger(__name__)
+
+def read_vectorized(collection, use_wkb=False):
+    cdef Session session
+    cdef void * cogr_feature
+    cdef void * cogr_geometry
+    cdef int num_fields
+    cdef void * fdefn
+    cdef int feature_index
+    cdef int field_index
+    cdef char * field_name_c
+    cdef bytes field_name_bytes
+    cdef int i
+    cdef int l
+    cdef int y = 0
+    cdef int m = 0
+    cdef int d = 0
+    cdef int hh = 0
+    cdef int mm = 0
+    cdef int ss = 0
+    cdef int tz = 0
+    cdef long long fid
+    cdef long long [:] arr_int
+    cdef double [:] arr_double
+    cdef char * wkt
+    cdef char * wkb
+
+    session = collection.session
+    encoding = session._fileencoding
+
+    if session.cogr_layer == NULL:
+        raise ValueError("Null layer")
+
+    length = OGR_L_GetFeatureCount(session.cogr_layer, 0)
+
+    data_fids = np.empty([length], dtype=object)
+    data_properties = {}
```

Comment:
Thoughts on using a [structured array](https://docs.scipy.org/doc/numpy-1.15.0/user/basics.rec.html) for `data_properties`?